File: native/android/app/src/main/java/com/mediconnect/nativeapp/AppointmentCancellation.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.coroutines.coroutineContext

interface CancellationService {
    suspend fun find(identity: Identity, appointmentId: String): Appointment
    suspend fun cancel(identity: Identity, appointmentId: String)
}
enum class CancellationStep { CLOSED, CONFIRM, PROCESSING, CONFIRMED, UNCONFIRMED }
data class CancellationState(val step: CancellationStep = CancellationStep.CLOSED, val appointment: Appointment? = null)

class AppointmentCancellation(private val service: CancellationService?, private val policy: CancellationContract?,
                              private val scope: CoroutineScope) {
    private val mutable = MutableStateFlow(CancellationState())
    val state = mutable.asStateFlow()
    private var identity: Identity? = null
    private var operation: Job? = null
    private var generation = 0
    private val attempted = mutableSetOf<Pair<String, String>>()

    fun close(clearSession: Boolean = false) {
        generation++; operation?.cancel(); identity = null; mutable.value = CancellationState()
        if (clearSession) attempted.clear()
    }
    fun open(identity: Identity, appointment: Appointment) {
        if (service == null || policy == null || identity.role != Role.PATIENT || appointment.id.isBlank()) return
        close(); this.identity = identity
        mutable.value = CancellationState(
            if (attempted.contains(identity.subject to appointment.id)) CancellationStep.UNCONFIRMED else CancellationStep.CONFIRM, appointment)
    }
    fun confirm() { if (mutable.value.step == CancellationStep.CONFIRM) run(submit = true) }
    fun check() { if (mutable.value.step == CancellationStep.UNCONFIRMED) run(submit = false) }
    private fun run(submit: Boolean) {
        val service = service ?: return
        val policy = policy ?: return
        val identity = identity ?: return
        val selected = mutable.value.appointment ?: return
        val current = ++generation
        mutable.value = CancellationState(CancellationStep.PROCESSING, selected)
        operation = scope.launch {
            try {
                var latest = service.find(identity, selected.id)
                coroutineContext.ensureActive()
                require(latest.id == selected.id)
                if (latest.status !in policy.cancelledStatuses && submit) {
                    require(latest.status in policy.cancellableStatuses)
                    require(attempted.add(identity.subject to selected.id))
                    service.cancel(identity, selected.id)
                    coroutineContext.ensureActive()
                    latest = service.find(identity, selected.id)
                    coroutineContext.ensureActive()
                }
                require(latest.id == selected.id && latest.status in policy.cancelledStatuses)
                if (current == generation) mutable.value = CancellationState(CancellationStep.CONFIRMED, latest)
            } catch (cancelled: CancellationException) { throw cancelled }
            catch (_: Exception) {
                if (current == generation) mutable.value = CancellationState(CancellationStep.UNCONFIRMED, selected)
            }
        }
    }
}

File: native/ios/Sources/AppointmentCancellation.swift
import Combine
import Foundation

@MainActor protocol CancellationService {
    func find(identity: Identity, appointmentId: String) async throws -> Appointment
    func cancel(identity: Identity, appointmentId: String) async throws
}
enum CancellationStep: Hashable { case closed, confirm, processing, confirmed, unconfirmed }
struct CancellationState {
    var step: CancellationStep = .closed
    var appointment: Appointment?
}
@MainActor final class AppointmentCancellation: ObservableObject {
    @Published private(set) var state = CancellationState()
    private let service: CancellationService?
    private let policy: CancellationContract?
    private var identity: Identity?
    private var operation: Task<Void, Never>?
    private var generation = 0
    private var attempted: [String: Set<String>] = [:]
    init(service: CancellationService?, policy: CancellationContract?) { self.service = service; self.policy = policy }
    func close(clearSession: Bool = false) {
        generation += 1; operation?.cancel(); identity = nil; state = CancellationState()
        if clearSession { attempted.removeAll() }
    }
    func open(_ identity: Identity, appointment: Appointment) {
        guard service != nil, policy != nil, identity.role == .patient, !appointment.id.isEmpty else { return }
        close(); self.identity = identity
        state = CancellationState(step: attempted[identity.subject, default: []].contains(appointment.id) ? .unconfirmed : .confirm, appointment: appointment)
    }
    func confirm() { if state.step == .confirm { run(submit: true) } }
    func check() { if state.step == .unconfirmed { run(submit: false) } }
    private func run(submit: Bool) {
        guard let service, let policy, let identity, let selected = state.appointment else { return }
        generation += 1; let current = generation
        state = CancellationState(step: .processing, appointment: selected)
        operation = Task { [weak self] in
            do {
                var latest = try await service.find(identity: identity, appointmentId: selected.id)
                try Task.checkCancellation()
                guard latest.id == selected.id else { throw MobileFailure.response }
                if !policy.cancelledStatuses.contains(latest.status ?? ""), submit {
                    guard policy.cancellableStatuses.contains(latest.status ?? ""), let self,
                          attempted[identity.subject, default: []].insert(selected.id).inserted else { throw MobileFailure.forbidden }
                    try await service.cancel(identity: identity, appointmentId: selected.id)
                    try Task.checkCancellation()
                    latest = try await service.find(identity: identity, appointmentId: selected.id)
                    try Task.checkCancellation()
                }
                guard latest.id == selected.id, policy.cancelledStatuses.contains(latest.status ?? "") else { throw MobileFailure.response }
                if let self, current == generation { state = CancellationState(step: .confirmed, appointment: latest) }
            } catch is CancellationError { return }
            catch {
                if let self, current == generation, !Task.isCancelled { state = CancellationState(step: .unconfirmed, appointment: selected) }
            }
        }
    }
}
